packages/fetchai/connections/soef/connection.py

```diff
@@ -91,106 +147,349 @@ def __init__(
         self.unique_page_address = None  # type: Optional[str]
         self.agent_location = None  # type: Optional[Location]
         self.in_queue = None  # type: Optional[asyncio.Queue]
+        self._executor_pool: Optional[ThreadPoolExecutor] = None
+        self.chain_identifier: str = chain_identifier or "fetchai"
 
-    def send(self, envelope: Envelope) -> None:
+    @staticmethod
+    def _is_compatible_query(query: Query) -> bool:
         """
-        Send message handler.
+        Check if a query is compatible with the soef.
 
-        :param envelope: the envelope.
+        :param query: search query to check
+        :return: bool
+        """
+        constraints = [c for c in query.constraints if isinstance(c, Constraint)]
+        if len(constraints) == 0:  # pragma: nocover
+            return False
+
+        if ConstraintTypes.DISTANCE not in [
+            c.constraint_type.type for c in constraints
+        ]:  # pragma: nocover
+            return False
+
+        return True
+
+    @staticmethod
+    def _construct_personality_filter_params(
+        equality_constraints: List[Constraint],
+    ) -> Dict[str, List[str]]:
+        """
+        Construct a dictionary of personality filters.
+
+        :param equality_constraints: list of equality constraints
+        :return: bool
+        """
+        filters = []
+
+        for constraint in equality_constraints:
+            if constraint.attribute_name not in PERSONALITY_PIECES:
+                continue
+            filters.append(
+                constraint.attribute_name + "," + constraint.constraint_type.value
+            )
+        if not filters:  # pragma: nocover
+            return {}
+        return {"ppfilter": filters}
+
+    @staticmethod
+    def _construct_service_key_filter_params(
+        equality_constraints: List[Constraint],
+    ) -> Dict[str, List[str]]:
+        """
+        Construct a dictionary of service keys filters.
```

Comment:
Add a line: `We assume each equality constraint which is not a personality piece relates to a service key!"

Reply:
done